libdnf/base/goal.cpp

```diff
@@ -596,10 +604,136 @@ void Goal::Impl::add_rpms_to_goal(base::Transaction & transaction) {
                     settings.resolve_best(cfg_main),
                     settings.resolve_clean_requirements_on_remove());
                 break;
+            case GoalAction::REINSTALL: {
+                solv::IdQueue ids_nevra_installed;
+                for (auto id : ids) {
+                    rpm::PackageQuery query(installed);
+                    query.filter_nevra({pool.get_nevra(id)});
+                    if (query.empty()) {
+                        // Report when package with the same NEVRA is not installed
+                        transaction.p_impl->add_resolve_log(
+                            GoalAction::REINSTALL,
+                            GoalProblem::NOT_INSTALLED,
+                            settings,
+                            {pool.get_nevra(id)},
+                            {},
+                            false);
+                    } else {
+                        // Only installed packages can be reinstalled
+                        ids_nevra_installed.push_back(id);
+                    }
+                }
+                rpm_goal.add_install(
+                    ids_nevra_installed,
+                    settings.resolve_strict(cfg_main),
+                    settings.resolve_best(cfg_main),
+                    settings.resolve_clean_requirements_on_remove());
+            } break;
             case GoalAction::UPGRADE: {
+                solv::IdQueue ids_upgrades;
+                for (auto id : ids) {
+                    rpm::PackageQuery query(installed);
+                    query.filter_name({pool.get_name(id)});
+                    if (query.empty()) {
+                        // Report when package with the same name is not installed
+                        transaction.p_impl->add_resolve_log(
+                            GoalAction::UPGRADE,
+                            GoalProblem::NOT_INSTALLED,
+                            settings,
+                            {pool.get_nevra(id)},
+                            {},
+                            false);
+                    } else {
+                        query.filter_arch({pool.get_arch(id)});
```

Comment:
With arch it is not so easy. Specially with changes with arch `noarch` to arch or opposite direction.It can be that implementation was wrong but because it is rare no one reported the issue because most of users uses upgrade all.

Reply:
OK. I added support for from/to "noarch" conversion in upgrade.